libs/crosstarget-utils/src/wasm/regex.rs

```diff
@@ -0,0 +1,39 @@
+use js_sys::RegExp as JSRegExp;
+
+use crate::common::{RegExpError, RegExpFlags};
+
+pub struct RegExp {
+    inner: JSRegExp,
+}
+
+impl RegExp {
+    pub fn new(pattern: &str, flags: Vec<RegExpFlags>) -> Result<Self, RegExpError> {
+        let mut flags_as_str: String = flags.into_iter().map(|flag| String::from(flag)).collect();
+
+        // Global flag is implied in `regex::Regex`, so we match that behavior for consistency.
+        flags_as_str.push('g');
+
+        Ok(Self {
+            inner: JSRegExp::new(pattern, &flags_as_str),
+        })
+    }
+
+    /// Searches for the first match of this regex in the haystack given, and if found,
+    /// returns not only the overall match but also the matches of each capture group in the regex.
+    /// If no match is found, then None is returned.
+    pub fn captures(&self, message: &str) -> Option<Vec<String>> {
+        let matches = self.inner.exec(message);
+        matches.map(|matches| {
+            let mut captures = Vec::new();
+            for i in 0..matches.length() {
+                captures.push(matches.get(i).as_string().unwrap());
+            }
+            captures
+        })
```

Comment:
```suggestion
        self.inner.exec(message).map(|matches| {
            matches
                .iter()
                .map(|m| m.as_string().expect("match result should be a valid UTF-8 string"))
                .collect()
        })
```

Reply:
I just realised that these values may contain `undefined` as well ([reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec)). We ought to filter them out.

Reply:
I think you might be misreading the doc, the only thing regarding `undefined` I see there is that the additional `groups` property of the array object is `undefined` if there are no named groups — it's irrelevant to us since we don't use it and it doesn't influence the numeric keys in any way.

TypeScript also says `interface RegExpExecArray extends Array<string>`.

Am I missing something?

Reply:
Note also that if what you said is true, filtering `undefined` out is not a correct solution because it would mess up match group indexes, we can't change the length of the array without breaking the caller.

Reply:
Updated by keeping the same number of captures. Let's use `.unwrap_or_default()` rather than `.expect()`, as we really don't want to panic during a regex execution.

Reply:
fine with me, please see the new suggestion though